lib/icingadb/icingadb.cpp

```diff
@@ -47,6 +47,10 @@ void IcingaDB::Validate(int types, const ValidationUtils& utils)
 	if (!(types & FAConfig))
 		return;
 
+	if (!GetUsername().IsEmpty() && GetPassword().IsEmpty()) {
+		BOOST_THROW_EXCEPTION(ValidationError(this, std::vector<String>(), "Validation failed: If username is set, a password is required as well."));
+	}
```

Comment:
Actually, I don't think this is the right place to perform such a constraint and even the existing certificate path validation. The `IcingaDB::Validate()` method should validate the object as a whole, initialising the environment ID for example, not such a specific attribute validity check. Instead, you can put this in its own method by overriding the `IcingaDB::ValidateUsername()` method.

Reply:
Like this for example:
https://github.com/Icinga/icinga2/blob/8beb0b74ab9f8402abf4a05b9b768d312304e3f0/lib/icingadb/icingadb.cpp#L217-L224

Reply:
> Instead, you can put this in its own method by overriding the `IcingaDB::ValidateUsername()` method.

Aren't those for checking the values of individual fields in isolation?

> The `IcingaDB::Validate()` method should validate the object as a whole, initialising the environment ID for example

Initializing something in a validate method sounds more like a hack to me.

Reply:
I thought this was the right place due to the fact that the TLS verification just follows. But I can change it to a more specific validation method.

Reply:
> Aren't those for checking the values of individual fields in isolation?

As far as I'm concerned, `ValidateUsername()` should fail as well if this constraint isn't met, but I see that a few other types also perform this kind of validation in the base method, so just as you like it!

Reply:
> I'm not sure if this would even work: Would `GetPassword()` within `IcingaDB::ValidateUsername()` even return a value yet?

`lvalue` is literally a wrapper around a plain `GetFoo()` method call 😅.
```cxx
	if (2 & types)
		ValidateUsername(Lazy<String>([this]() { return GetUsername(); }), utils);
```

> Would it depend on the order the attributes are set within the object definition?

No, the attributes are all initialised at the validation stage.

Reply:
As it turns out, the individual validation methods are also called when modifying the attributes via the API:
https://github.com/Icinga/icinga2/blob/3ffec77ae52d7ea664edd57be7b6565f4f3b049d/lib/base/configobject.cpp#L193-L194

Since the `password` field is set to `no_user_modify`, meaning that it isn't modifiable via the API but with the current state of this PR, that sanity check wouldn't actually be triggered when changing/setting the `username` via the API.

However, looking at the current Icinga DB code, it doesn't seem to make sense to me to make the config attributes runtime modifiable at all, as once Icinga DB is started, it never uses any of those, i.e. you can e.g. change the Redis port at runtime, but Icinga DB won't notice that change. Therefore, I would just mark the `username` to `no_user_modify` as well to disallow changes at runtime.

Reply:
I am currently at reworking this, thus I would include your comment as well. Thanks for taking another look, @yhabteab.

Reply:
I force pushed some changes according to @yhabteab's comments, which can be verified via the API as follows:

- Trying to create an IcingaDB object with a username but without a password:
```
$ curl -k -s -S -u root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/IcingaDBs/test' -d '{"attrs": {"username": "foo"}, "pretty": true}'
{
    "results": [
        {
            "code": 500,
            "errors": [
                "Error: Validation failed for object 'test' of type 'IcingaDB'; Attribute 'username': If a username is given, a password is also required. However, a password can exist on its own.\nLocation: in /var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf: 2:2-2:17\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(1): object IcingaDB \"test\" {\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(2):  username = \"foo\"\n                                                                                                        ^^^^^^^^^^^^^^^^\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(3):  version = 1726755773.422595\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(4):  zone = \"master\"\n"
            ],
            "status": "Object could not be created."
        }
    ]
}
```
- Trying to create an IcingaDB object with a faulty user certificate config (only `cert_path` being set):
```
$ curl -k -s -S -u root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/IcingaDBs/test' -d '{"attrs": {"enable_tls": true, "cert_path": "/foo"}, "pretty": true}'
{
    "results": [
        {
            "code": 500,
            "errors": [
                "Error: Validation failed for object 'test' of type 'IcingaDB'; Attribute 'cert_path': Either both a client certificate (cert_path) and its private key (key_path) or none of them must be given.\nLocation: in /var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf: 2:2-2:19\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(1): object IcingaDB \"test\" {\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(2):  cert_path = \"/foo\"\n                                                                                                        ^^^^^^^^^^^^^^^^^^\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(3):  enable_tls = true\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(4):  version = 1726755898.696290\n"
            ],
            "status": "Object could not be created."
        }
    ]
}
```
- Trying to create an IcingaDB object with a faulty user certificate config (only `key_path` being set):
```
$ curl -k -s -S -u root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/IcingaDBs/test' -d '{"attrs": {"enable_tls": true, "key_path": "/foo"}, "pretty": true}'
{
    "results": [
        {
            "code": 500,
            "errors": [
                "Error: Validation failed for object 'test' of type 'IcingaDB'; Attribute 'cert_path': Either both a client certificate (cert_path) and its private key (key_path) or none of them must be given.\nLocation: in /var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf: 1:0-1:21\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(1): object IcingaDB \"test\" {\n                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(2):  enable_tls = true\n/var/lib/icinga2/api/packages/_api/c9ef1947-e1ff-47f1-9afa-385ad39c88a2/conf.d/icingadbs/test.conf(3):  key_path = \"/foo\"\n"
            ],
            "status": "Object could not be created."
        }
    ]
}
```
- Successfully creating an IcingaDB object with username and password:
```
$ curl -k -s -S -u root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/IcingaDBs/test' -d '{"attrs": {"username": "foo", "password": "bar"}, "pretty": true}'
{
    "results": [
        {
            "code": 200,
            "status": "Object was created"
        }
    ]
}
```
- Trying to change the username of the just created IcingaDB object, which should fail now:
```
$ curl -k -s -S -u root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/objects/IcingaDBs/test' -d '{"attrs": {"username": "change"}, "pretty": true}'

{
    "results": [
        {
            "code": 500,
            "name": "test",
            "status": "Attribute 'username' could not be set: Error: Attribute cannot be modified.\n",
            "type": "IcingaDB"
        }
    ]
}
```